lib/Service/MailManager.php

```diff
@@ -419,6 +427,50 @@ public function flagMessage(Account $account, string $mailbox, int $uid, string
 		);
 	}
 
+	/**
+	 * Tag (flag) a message on IMAP
+	 *
+	 * @param Account $account
+	 * @param string $mailbox
+	 * @param integer $uid
+	 * @param array $tag
+	 * @param boolean $value
+	 * @return void
+	 *
+	 * @uses
+	 *
+	 * @link https://github.com/nextcloud/mail/issues/25
+	 */
+	public function tagMessage(Account $account, string $mailbox, Message $message, Tag $tag, bool $value): void {
+		$client = $this->imapClientFactory->getClient($account);
+		try {
+			$mb = $this->mailboxMapper->find($account, $mailbox);
+		} catch (DoesNotExistException $e) {
+			throw new ClientException("Mailbox $mailbox does not exist", 0, $e);
+		}
+		if ($this->isPermflagsEnabled($account, $mailbox) === true) {
+			try {
+				if ($value === true) {
+					// imap keywords and flags work the same way
+					$this->imapMessageMapper->addFlag($client, $mb, $message->getUid(), $tag->getImapLabel());
+				} else {
+					$this->imapMessageMapper->removeFlag($client, $mb, $message->getUid(), $tag->getImapLabel());
+				}
+			} catch (Horde_Imap_Client_Exception $e) {
+				throw new ServiceException(
+					"Could not set message keyword on IMAP: " . $e->getMessage(),
+					(int) $e->getCode(),
+					$e
+				);
+			}
+		}
+		if ($value === true) {
```

Comment:
```suggestion
		if ($value) {
```

it's a bool as per typehint